arangod/Agency/TransactionBuilder.h

```diff
@@ -33,210 +33,234 @@
-    _buffer.openObject();
-    return std::move(_buffer);
-  }
-  write_trx& operator=(write_trx&&) = default;
-
- private:
-  friend T;
-
-  explicit write_trx(buffer &&b) : _buffer(std::move(b)) {}
-  buffer _buffer;
-};
-
-
-template<typename B>
-struct envelope {
-  using buffer = buffer_mapper<B>;
-
-  read_trx<B, envelope> read() && {
-    _buffer.openArray();
-    return read_trx<B, envelope>(std::move(_buffer));
-  }
-
-  write_trx<B, envelope> write() && {
-    _buffer.openArray();
-    _buffer.openObject();
-    return write_trx<B, envelope>(std::move(_buffer));
+namespace detail {
+template <typename T>
+struct moving_ptr {
+  explicit moving_ptr(T* p) : _ptr(p) {}
+  ~moving_ptr() noexcept = default;
+
+  moving_ptr& operator=(moving_ptr const&) = delete;
+  moving_ptr(moving_ptr const&) = delete;
+
+  moving_ptr(moving_ptr&& o) { *this = std::move(o); }
+  moving_ptr& operator=(moving_ptr&& o) {
+    reset(o.release());
+    return *this;
   }
 
-  void done() && {
-    _buffer.closeArray();
+  T* release() noexcept {
+    T* result = _ptr;
+    _ptr = nullptr;
+    return result;
   }
+  void reset(T* p) { _ptr = p; }
 
-  static envelope create(VPackBuilder &b) {
-    buffer buff(b);
-    envelope env(buffer{b});
-    env._buffer.openArray();
-    return env;
-  }
+  T* operator->() { return _ptr; }
+  T* get() { return _ptr; }
+  T& operator*() { return *_ptr; }
+  operator bool() { return _ptr != nullptr; }
 
  private:
-  friend write_trx<B, envelope>;
-  friend read_trx<B, envelope>;
-  friend precs_trx<B, envelope>;
-  explicit envelope(buffer&& b) : _buffer(std::move(b)) {}
-  envelope() = default;
-  buffer _buffer;
+  T* _ptr = nullptr;
 };
 
-namespace detail {
-
-template<typename V>
+template <typename V>
 void add_to_builder(VPackBuilder* b, V const& v) {
   b->add(VPackValue(v));
 }
 
-template<>
+template <>
 inline void add_to_builder(VPackBuilder* b, VPackSlice const& v) {
   b->add(v);
 }
 
+template <typename K, typename V>
+void add_to_builder(VPackBuilder* b, K const& key, V const& v) {
+  b->add(key, VPackValue(v));
 }
 
+template <typename K>
+inline void add_to_builder(VPackBuilder* b, K const& key, VPackSlice const& v) {
+  b->add(key, v);
+}
+}  // namespace detail
 
-template<>
-struct buffer_mapper<VPackBuilder> {
-
-  explicit buffer_mapper(VPackBuilder& builder) : _builder(&builder) {};
-
-  template<typename K, typename V>
-  void setKey(K&& k, V&& v) {
-      setValue(std::forward<K>(k));
-      setValue(std::forward<V>(v));
+struct envelope {
+  using builder_ptr = detail::moving_ptr<VPackBuilder>;
+
+  struct read_trx {
+    envelope end() {
+      _builder->close();
+      return envelope(_builder.release());
+    }
+    template <typename K>
+    read_trx& key(K&& k) {
+      detail::add_to_builder(_builder.get(), std::forward<K>(k));
+      return *this;
+    }
+    ~read_trx() noexcept {
+      try {
+        if (_builder) {
+          end();
+        }
+      } catch (...) {
+      }
+    }
```

Comment:
I'm unsure whether it's a good idea to try to silently repair when a user forgot to call `end()`. That might well break the velocypack in some unexpected way, as it will probably not be called at the right time.

There could even an argument to be made for throwing an exception if the builder is not released on destruction, though that may be bad for other reasons. Maybe in maintainer mode?

Please share your thoughts. :)

Reply:
I would prefer to always crash, even in release build. However, I think some people in the company disagree on that. So I could add an assertion for Maintainer mode and we ignore things in release mode :man_shrugging:

Reply:
Well, maybe not always crash, because there is out of memory exception and this should be able to bubble up.